### PR TITLE
feat(CI): Added GitHub Actions to check for code style violations

### LINKED ISSUE #13413: CI: Add Auto-Format Workflow for Java files
- [x] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.
- [x] I have searched the [release notes](https://github.com/apache/dubbo/releases) of this repository and believe that this is not a duplicate.

## Describe the feature

This project has a code style available in the `codestyle` directory. Checking whether the code in the Pull Requests follows the project's code style and applying fixes accordingly is a tedious work. **I want to automate this check and it will also fix any code that violates the code style. The changes will be pushed to the source branch of the PR automatically.** This check can also be applied for any branch of this repository.
**I would like to work on this issue.**

### PR DESCRIPTION
**Pull request by SaptarshiSarkar12**
## What is the purpose of the change
Closes #13413 

## Brief changelog
- Added GitHub Actions to check for code style violations.

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [X] Each commit in the pull request should have a meaningful subject line and body.
- [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [X] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ I have added a slight tab to break the style of one of the java file to check if the CI actually works.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ I think the workflows don't have **read and write permission**. You can go to `Settings` > `Actions` > `general` and turn on this option :point_down: 
image

**Comment by SaptarshiSarkar12** [comment]
> @AlbumenJ I think the workflows don't have **read and write permission**. You can go to `Settings` > `Actions` > `general` and turn on this option 👇 image

@AlbumenJ The workflow will fail unless you set these permissions.

**Comment by AlbumenJ** [comment]
> > @AlbumenJ I think the workflows don't have **read and write permission**. You can go to `Settings` > `Actions` > `general` and turn on this option 👇 image
> 
> @AlbumenJ The workflow will fail unless you set these permissions.

Seems it should be changed to scheduled task to scan PRs due to permission limitation. The ASF will not enable write access in PRs due to workflows file can be changed by anyone. So it might be a good idea to switch to other kinds of task.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ Okay, I see. So, do you mean that this workflow should run on schedule and will just check if codestyle is followed?

**Comment by AlbumenJ** [comment]
> @AlbumenJ Okay, I see. So, do you mean that this workflow should run on schedule and will just check if codestyle is followed?

Yes :)

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ Okay. Then, what should be the time of the workflow run in UTC timezone (this timezone is followed by GH Actions)? Should I keep it at midnight UTC everyday? You can check and give me a custom cron time from https://crontab.guru/ .

**Comment by AlbumenJ** [comment]
> @AlbumenJ Okay. Then, what should be the time of the workflow run in UTC timezone (this timezone is followed by GH Actions)? Should I keep it at midnight UTC everyday? You can check and give me a custom cron time from https://crontab.guru/ .

Every hour or even 20 mins are both ok

**Comment by SaptarshiSarkar12** [comment]
Okay @AlbumenJ

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ I have added the cron job in the workflow to run at every 20 mins.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ The cron job worked :tada: - you can check here - https://github.com/SaptarshiSarkar12/dubbo/actions/runs/7330812949

**Comment by CrazyHZM** [comment]
@SaptarshiSarkar12 
When PR is submitted, the existing ci process will automatically execute `mvn spotless check` due to the execution of `mvn install`. I can't think of the function of this schedule. The code format is guaranteed because this check is done for every commit.

I have an individual suggestion: add a ci code checkstyle to each submission instead of completing it through schedule.

**Comment by SaptarshiSarkar12** [comment]
@CrazyHZM Okay. But, I couldn't understand your suggestion. Can you please elaborate?

**Comment by CrazyHZM** [comment]
Dubbo has 18 ci tasks, and the code formatted check isn't standalone. It misleads developers to understand why ci fails.
We can add a ci task for code formated check standalone.
@SaptarshiSarkar12 

ci

**Comment by SaptarshiSarkar12** [comment]
@CrazyHZM So, you mean I will create a CI task that check only the formatting? Like running a goal of the maven plugin?

**Comment by CrazyHZM** [comment]
> @CrazyHZM So, you mean I will create a CI task that check only the formatting? Like running a goal of the maven plugin?

Running` maven spotless:check ` should be enough.

**Comment by SaptarshiSarkar12** [comment]
@CrazyHZM I have already created the workflow which runs that command. You can check [here](https://github.com/apache/dubbo/pull/13489/files#diff-173f5db567f8459284c719180df8c27debad26456e63f98c5b87d94cf87e049bR20).
Do you mean to change that workflow from **schedule** to **on push** (just as other workflow)?

**Comment by CrazyHZM** [comment]
> @CrazyHZM I have already created the workflow which runs that command. You can check [here](https://github.com/apache/dubbo/pull/13489/files#diff-173f5db567f8459284c719180df8c27debad26456e63f98c5b87d94cf87e049bR20). Do you mean to change that workflow from **schedule** to **on push** (just as other workflow)?

Yes.

**Comment by SaptarshiSarkar12** [comment]
Ok. I'll do that. Thank you for the suggestion @CrazyHZM 😁!

**Comment by SaptarshiSarkar12** [comment]
@CrazyHZM I have made the required changes.
@AlbumenJ @CrazyHZM Please let me know if any other changes are required.
Thank you!

File: .github/workflows/check-format.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Setup Maven
+        uses: actions/setup-java@v4
+        with:
+          java-version: 21
+          distribution: zulu
+      - name: Check if code aligns with code style
+        run: mvn spotless:check
```
Comment (AlbumenJ) [inline .github/workflows/check-format.yml:20]: Can we just log out something to the `Workflow Summary`?
Like this
image


After that, we can make this job as the first job for pull requests and make sure that no other message will influence contributors.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: .github/workflows/check-format.yml (hunk 1 above)
Comment (SaptarshiSarkar12) [reply .github/workflows/check-format.yml:20]: Yes, there's an option. I'll try it locally. If it works, then we'll and good 👍. Else, need to find some other workaround.

Comment (SaptarshiSarkar12) [reply .github/workflows/check-format.yml:20]: @AlbumenJ I'll let you know 🙂.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ I think this is the summary that you were talking about, isn't it? I think the CI should also fail if the maven plugin has failed, what do you think?

**Comment by AlbumenJ** [comment]
> @AlbumenJ I think [this is the summary](https://github.com/apache/dubbo/actions/runs/7413158276/attempts/2#summary-20171547993) that you were talking about, isn't it? I think the CI should also fail if the maven plugin has failed, what do you think?

Great! And it's a good idea to add some prominent reminders to remind contributors how to fix it.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ Sorry, I couldn't understand your words. What did you mean by **reminder**? Should I add a line at the end of the summary with the correct command to fix the format error?

**Comment by AlbumenJ** [comment]
> @AlbumenJ Sorry, I couldn't understand your words. What did you mean by **reminder**? Should I add a line at the end of the summary with the correct command to fix the format error?

Or at the first

**Comment by SaptarshiSarkar12** [comment]
Okay got it @AlbumenJ.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ I have made the required changes. Please review it.

**Comment by AlbumenJ** [comment]
1. How about set `Check if code needs formatting` as the required step for all the steps in `Build and Test For PR`. This means we should move `Check for formatting` into `Build and Test For PR`.
2. Why use `--log-file mvn.log` in `Check if code aligns with code style`

Others LGMT

**Comment by SaptarshiSarkar12** [comment]
> 1. How about set `Check if code needs formatting` as the required step for all the steps in `Build and Test For PR`. This means we should move `Check for formatting` into `Build and Test For PR`.

@AlbumenJ Yeah, that's a good idea.
> 2. Why use `--log-file mvn.log` in `Check if code aligns with code style`

@AlbumenJ I have tried using `gh run view --log` command to get the log, but it does not work until the run completes its execution. 
So, I have used that arg to dump the maven execution output to a log file from which I'll extract the error and show it in the Workflow summary.

**Comment by AlbumenJ** [comment]
> @AlbumenJ I have tried using `gh run view --log` command to get the log, but it does not work until the run completes its execution. So, I have used that arg to dump the maven execution output to a log file from which I'll extract the error and show it in the Workflow summary.

Print all the logs in a separate step would be better? Or upload "mvn.log" as an artifact. If so, we can determine if other Maven plugins are working as expected if the workflow fails due to other issues.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ Yeah, that's a good idea. I prefer uploading it as an artifact so that it can be downloaded and shared easily fostering more collaboration. What do you think?

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ Should I add the **Check for formatting** workflow to **Build and Test Schedule** also?

**Comment by AlbumenJ** [comment]
> @AlbumenJ Should I add the **Check for formatting** workflow to **Build and Test Schedule** also?

I think it can be omitted. Because the scheduled task itself will check the format.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ Should the workflow remain in the Build and test PR workflow? I couldn't understand your words. Can you please elaborate?

**Comment by AlbumenJ** [comment]
> @AlbumenJ Should the workflow remain in the Build and test PR workflow? I couldn't understand your words. Can you please elaborate?

Sorry about that. The workflow of `Build and Test Schedule` has already contained the check the code style with `spotless` now. 
Maybe in https://github.com/apache/dubbo/blob/05d44abe933f03024f57efd8779d6595f38e7395/.github/workflows/build-and-test-scheduled-3.2.yml#L167

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ It's okay.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ Please review this PR then. Let me know if any further changes are required.

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ I don't think the **scheduled check** creates summary if linting fails. Can I try adding it?

**Comment by AlbumenJ** [comment]
> @AlbumenJ Please review this PR then. Let me know if any further changes are required.

This PR LGTM

**Comment by AlbumenJ** [comment]
> @AlbumenJ I don't think the **scheduled check** creates summary if linting fails. Can I try adding it?

`Build and Test For PR` is enough now. **scheduled check** is for maintainers.

**Comment by SaptarshiSarkar12** [comment]
Okay @AlbumenJ. So, can I remove my changes which were made solely for testing the CI (like adding extra spaces)?

**Comment by AlbumenJ** [comment]
> Okay @AlbumenJ. So, can I remove my changes which were made solely for testing the CI (like adding extra spaces)?

Yep

**Comment by SaptarshiSarkar12** [comment]
Okay @AlbumenJ. Would do it soon.
Thank you!

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ I have made the required changes.

**Review by AlbumenJ — APPROVED** [review]
LGTM

**Comment by AlbumenJ** [comment]
@CrazyHZM PTAL

**Review by CrazyHZM — APPROVED** [review]

**Comment by AlbumenJ** [comment]
@SaptarshiSarkar12 Thanks for your contribution :)

**Comment by SaptarshiSarkar12** [comment]
@AlbumenJ I have a question. What is the full form / meaning of PTAL?
@AlbumenJ @CrazyHZM Thank you for reviewing and merging my PR :smile:! I loved working on this project :heart:.

**Comment by refeccd** [comment]
`PTAL` Please take a look

**Comment by SaptarshiSarkar12** [comment]
Okay. Got it. Thank you @liaozan :smile:!

**Pull request merged**